### PR TITLE
Add psalm linter for PHP

### PR DESCRIPTION
**Pull request by twitchard**
Submitted this previously in #1593, but I didn't follow through with the requested changes and it was closed for inactivity. This time, I added the documentation.

### CODE REVIEW

File: ale_linters/php/psalm.vim [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+    return ale#Var(a:buffer, 'php_psalm_executable')
+endfunction
+
+function! ale_linters#php#psalm#GetCommand(buffer) abort
+    return ale#Escape(ale_linters#php#psalm#GetExecutable(a:buffer)) . ' --diff --output-format=emacs %s'
+endfunction
+
+function! ale_linters#php#psalm#Handle(buffer, lines) abort
+    " Matches patterns like the following:
+
```
Comment (w0rp) [inline ale_linters/php/psalm.vim]: Remove the blank line here.

File: ale_linters/php/psalm.vim [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+" Description: plugin for Psalm, static analyzer for PHP
+
+let g:ale_php_psalm_executable = get(g:, 'ale_php_psalm_executable', 'psalm')
+
+function! ale_linters#php#psalm#GetExecutable(buffer) abort
+    return ale#Var(a:buffer, 'php_psalm_executable')
+endfunction
+
+function! ale_linters#php#psalm#GetCommand(buffer) abort
+    return ale#Escape(ale_linters#php#psalm#GetExecutable(a:buffer)) . ' --diff --output-format=emacs %s'
```
Comment (w0rp) [inline ale_linters/php/psalm.vim]: You can write `command: '%e --diff --output-format=emacs %s`' below instead, and get the same result.

File: ale_linters/php/psalm.vim [hunk 3]
```diff
@@ -0,0 +1,38 @@
+" Author: richard marmorstein <https://github.com/twitchard>
+" Description: plugin for Psalm, static analyzer for PHP
+
+let g:ale_php_psalm_executable = get(g:, 'ale_php_psalm_executable', 'psalm')
+
+function! ale_linters#php#psalm#GetExecutable(buffer) abort
+    return ale#Var(a:buffer, 'php_psalm_executable')
```
Comment (w0rp) [inline ale_linters/php/psalm.vim]: You can write `'executable_callback': ale#VarFunc('php_psalm_executable')` below instead.

File: ale_linters/php/psalm.vim [hunk 4]
```diff
@@ -0,0 +1,38 @@
+" Author: richard marmorstein <https://github.com/twitchard>
+" Description: plugin for Psalm, static analyzer for PHP
+
+let g:ale_php_psalm_executable = get(g:, 'ale_php_psalm_executable', 'psalm')
```
Comment (w0rp) [inline ale_linters/php/psalm.vim]: You can write `call ale#Set('php_psalm_executable', 'psalm')` instead.

File: doc/ale-php.txt [hunk 5]
```diff
@@ -169,6 +169,16 @@ g:ale_php_phpstan_configuration               *g:ale_php_phpstan_configuration*
   This variable sets path to phpstan configuration file.
 
 
+===============================================================================
+psalm                                                           *ale-php-psalm*
```
Comment (w0rp) [inline doc/ale-php.txt]: Update the table of contents to include this heading too.

File: ale_linters/php/psalm.vim [hunk 6]
```diff
@@ -0,0 +1,38 @@
+" Author: richard marmorstein <https://github.com/twitchard>
```
Comment (w0rp) [inline ale_linters/php/psalm.vim]: Add some tests for the command and executable here. See `:help ale-development`.

**Comment by twitchard** [comment]
All right! I believe I've addressed all your comments.

I really admire your project by the way. The documentation and tooling around contributing is extremely impressive.

**Comment by twitchard** [comment]
Anything else before this is ready to merge, @w0rp ?

**Pull request merged**